File: Python/OneLang/One/Transforms/InstanceOfImplicitCast.py
from onelang_core import *
import OneLang.One.AstTransformer as astTrans
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.Statements as stats
import OneLang.One.Ast.References as refs
import OneLang.Utils.ArrayHelper as arrayHelp

class InstanceOfImplicitCast(astTrans.AstTransformer):
    def __init__(self):
        self.casts = []
        self.cast_counts = []
        super().__init__("InstanceOfImplicitCast")
    
    def add_cast(self, cast):
        if len(self.cast_counts) > 0:
            cast.implicit_casts = []
            self.casts.append(cast)
            last = len(self.cast_counts) - 1
            self.cast_counts[last] = self.cast_counts[last] + 1
    
    def push_context(self):
        self.cast_counts.append(0)
    
    def pop_context(self):
        cast_count = self.cast_counts.pop()
        if cast_count != 0:
            del self.casts[-cast_count:]
    
    def equals(self, expr1, expr2):
        # implicit casts don't matter when checking equality...
        while isinstance(expr1, exprs.CastExpression) and expr1.instance_of_cast != None:
            expr1 = expr1.expression
        while isinstance(expr2, exprs.CastExpression) and expr2.instance_of_cast != None:
            expr2 = expr2.expression
        
        # MetP, V, MethP.PA, V.PA, MethP/V [ {FEVR} ], FEVR
        if isinstance(expr1, exprs.PropertyAccessExpression):
            return isinstance(expr2, exprs.PropertyAccessExpression) and expr1.property_name == expr2.property_name and self.equals(expr1.object, expr2.object)
        elif isinstance(expr1, refs.VariableDeclarationReference):
            return isinstance(expr2, refs.VariableDeclarationReference) and expr1.decl == expr2.decl
        elif isinstance(expr1, refs.MethodParameterReference):
            return isinstance(expr2, refs.MethodParameterReference) and expr1.decl == expr2.decl
        elif isinstance(expr1, refs.ForeachVariableReference):
            return isinstance(expr2, refs.ForeachVariableReference) and expr1.decl == expr2.decl
        elif isinstance(expr1, refs.InstanceFieldReference):
            return isinstance(expr2, refs.InstanceFieldReference) and expr1.field == expr2.field
        elif isinstance(expr1, refs.ThisReference):
            return isinstance(expr2, refs.ThisReference)
        elif isinstance(expr1, refs.StaticThisReference):
            return isinstance(expr2, refs.StaticThisReference)
        return False
    
    def visit_expression(self, expr):
        result = expr
        if isinstance(expr, exprs.InstanceOfExpression):
            self.visit_expression(expr.expr)
            self.add_cast(expr)
        elif isinstance(expr, exprs.BinaryExpression) and expr.operator == "&&":
            expr.left = self.visit_expression(expr.left)
            expr.right = self.visit_expression(expr.right)
        elif isinstance(expr, exprs.ConditionalExpression):
            self.push_context()
            expr.condition = self.visit_expression(expr.condition)
            expr.when_true = self.visit_expression(expr.when_true)
            self.pop_context()
            
            expr.when_false = self.visit_expression(expr.when_false)
        elif isinstance(expr, refs.Reference) and isinstance(expr.parent_node, exprs.BinaryExpression) and expr.parent_node.operator == "=" and expr.parent_node.left == expr:
            pass
        else:
            self.push_context()
            result = super().visit_expression(expr)
            self.pop_context()
            # @java final var result2 = result;
            result2 = result
            match = next(filter(lambda cast: self.equals(result2, cast.expr), self.casts), None)
            if match != None:
                cast_expr = exprs.CastExpression(match.check_type, result)
                cast_expr.instance_of_cast = match
                match.implicit_casts.append(cast_expr)
                result = cast_expr
        return result
    
    def visit_statement(self, stmt):
        self.current_statement = stmt
        
        if isinstance(stmt, stats.IfStatement):
            self.push_context()
            stmt.condition = self.visit_expression(stmt.condition)
            self.visit_block(stmt.then)
            self.pop_context()
            
            if stmt.else_ != None:
                self.visit_block(stmt.else_)
        elif isinstance(stmt, stats.WhileStatement):
            self.push_context()
            stmt.condition = self.visit_expression(stmt.condition)
            self.visit_block(stmt.body)
            self.pop_context()
        else:
            self.push_context()
            super().visit_statement(stmt)
            self.pop_context()
        
        return stmt

File: PHP/src/OneLang/One/Transforms/InstanceOfImplicitCast.php
<?php

namespace OneLang\One\Transforms\InstanceOfImplicitCast;

use OneLang\One\AstTransformer\AstTransformer;
use OneLang\One\Ast\Expressions\InstanceOfExpression;
use OneLang\One\Ast\Expressions\BinaryExpression;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\CastExpression;
use OneLang\One\Ast\Expressions\ConditionalExpression;
use OneLang\One\Ast\Expressions\PropertyAccessExpression;
use OneLang\One\Ast\Statements\Statement;
use OneLang\One\Ast\Statements\IfStatement;
use OneLang\One\Ast\Statements\WhileStatement;
use OneLang\One\Ast\References\ForeachVariableReference;
use OneLang\One\Ast\References\VariableDeclarationReference;
use OneLang\One\Ast\References\MethodParameterReference;
use OneLang\One\Ast\References\InstanceFieldReference;
use OneLang\One\Ast\References\ThisReference;
use OneLang\One\Ast\References\Reference;
use OneLang\One\Ast\References\StaticThisReference;
use OneLang\Utils\ArrayHelper\ArrayHelper;

class InstanceOfImplicitCast extends AstTransformer {
    public $casts;
    public $castCounts;
    
    function __construct() {
        parent::__construct("InstanceOfImplicitCast");
        $this->casts = array();
        $this->castCounts = array();
    }
    
    protected function addCast($cast) {
        if (count($this->castCounts) > 0) {
            $cast->implicitCasts = array();
            $this->casts[] = $cast;
            $last = count($this->castCounts) - 1;
            $this->castCounts[$last] = $this->castCounts[$last] + 1;
        }
    }
    
    protected function pushContext() {
        $this->castCounts[] = 0;
    }
    
    protected function popContext() {
        $castCount = array_pop($this->castCounts);
        if ($castCount !== 0)
            array_splice($this->casts, -$castCount);
    }
    
    protected function equals($expr1, $expr2) {
        // implicit casts don't matter when checking equality...
        while ($expr1 instanceof CastExpression && $expr1->instanceOfCast !== null)
            $expr1 = $expr1->expression;
        while ($expr2 instanceof CastExpression && $expr2->instanceOfCast !== null)
            $expr2 = $expr2->expression;
        
        // MetP, V, MethP.PA, V.PA, MethP/V [ {FEVR} ], FEVR
        if ($expr1 instanceof PropertyAccessExpression)
            return $expr2 instanceof PropertyAccessExpression && $expr1->propertyName === $expr2->propertyName && $this->equals($expr1->object, $expr2->object);
        else if ($expr1 instanceof VariableDeclarationReference)
            return $expr2 instanceof VariableDeclarationReference && $expr1->decl === $expr2->decl;
        else if ($expr1 instanceof MethodParameterReference)
            return $expr2 instanceof MethodParameterReference && $expr1->decl === $expr2->decl;
        else if ($expr1 instanceof ForeachVariableReference)
            return $expr2 instanceof ForeachVariableReference && $expr1->decl === $expr2->decl;
        else if ($expr1 instanceof InstanceFieldReference)
            return $expr2 instanceof InstanceFieldReference && $expr1->field === $expr2->field;
        else if ($expr1 instanceof ThisReference)
            return $expr2 instanceof ThisReference;
        else if ($expr1 instanceof StaticThisReference)
            return $expr2 instanceof StaticThisReference;
        return false;
    }
    
    protected function visitExpression($expr) {
        $result = $expr;
        if ($expr instanceof InstanceOfExpression) {
            $this->visitExpression($expr->expr);
            $this->addCast($expr);
        }
        else if ($expr instanceof BinaryExpression && $expr->operator === "&&") {
            $expr->left = $this->visitExpression($expr->left);
            $expr->right = $this->visitExpression($expr->right);
        }
        else if ($expr instanceof ConditionalExpression) {
            $this->pushContext();
            $expr->condition = $this->visitExpression($expr->condition);
            $expr->whenTrue = $this->visitExpression($expr->whenTrue);
            $this->popContext();
            
            $expr->whenFalse = $this->visitExpression($expr->whenFalse);
        }
        else if ($expr instanceof Reference && $expr->parentNode instanceof BinaryExpression && $expr->parentNode->operator === "=" && $expr->parentNode->left === $expr) { }
        else {
            $this->pushContext();
            $result = parent::visitExpression($expr);
            $this->popContext();
            // @java final var result2 = result;
            $result2 = $result;
            $match = \OneLang\Core\ArrayHelper::find($this->casts, function ($cast) use ($result2) { return $this->equals($result2, $cast->expr); });
            if ($match !== null) {
                $castExpr = new CastExpression($match->checkType, $result);
                $castExpr->instanceOfCast = $match;
                $match->implicitCasts[] = $castExpr;
                $result = $castExpr;
            }
        }
        return $result;
    }
    
    protected function visitStatement($stmt) {
        $this->currentStatement = $stmt;
        
        if ($stmt instanceof IfStatement) {
            $this->pushContext();
            $stmt->condition = $this->visitExpression($stmt->condition);
            $this->visitBlock($stmt->then);
            $this->popContext();
            
            if ($stmt->else_ !== null)
                $this->visitBlock($stmt->else_);
        }
        else if ($stmt instanceof WhileStatement) {
            $this->pushContext();
            $stmt->condition = $this->visitExpression($stmt->condition);
            $this->visitBlock($stmt->body);
            $this->popContext();
        }
        else {
            $this->pushContext();
            parent::visitStatement($stmt);
            $this->popContext();
        }
        
        return $stmt;
    }
}
